lepton/image.go

```diff
@@ -249,62 +256,82 @@ func (bc dummy) Write(p []byte) (int, error) {
 	return len(p), nil
 }
 
-func DownloadBootImages() error {
-	return DownloadImages(ReleaseBaseUrl, false)
+func DownloadBootImages(baseUrl string, force bool) error {
+	home, err := HomeDir()
+	if err != nil {
+		return errors.Wrap(err, 1)
+	}
+	opsDirectory := path.Join(home, OpsDir)
+	if _, err := os.Stat(opsDirectory); os.IsNotExist(err) {
+		if err := os.MkdirAll(opsDirectory, 0755); err != nil {
+			return errors.Wrap(err, 1)
+		}
+	}
+	if baseUrl == "" {
+		return DownloadImages(ReleaseBaseUrl, false, opsDirectory)
+	}
+	return DownloadImages(baseUrl, force, opsDirectory)
 }
 
 // DownloadImages downloads latest kernel images.
-func DownloadImages(baseUrl string, force bool) error {
-	var err error
-	if _, err := os.Stat(".staging"); os.IsNotExist(err) {
-		os.MkdirAll(".staging", 0755)
-	}
+func DownloadImages(baseUrl string, force bool, baseDir string) error {
+
+	mkfsFilePath := path.Join(baseDir, Mkfs)
+	bootImgFilePath := path.Join(baseDir, BootImg)
+	kernelImgFilePath := path.Join(baseDir, KernelImg)
 
-	if _, err = os.Stat(Mkfs); os.IsNotExist(err) || force {
-		if err = DownloadFile(Mkfs, fmt.Sprintf(baseUrl, path.Join(runtime.GOOS, "mkfs")), 600); err != nil {
+	if _, err := os.Stat(mkfsFilePath); os.IsNotExist(err) || force {
+		if err = DownloadFile(mkfsFilePath, fmt.Sprintf(baseUrl, path.Join(runtime.GOOS, "mkfs")), 600); err != nil {
 			return errors.Wrap(err, 1)
 		}
 	}
 
 	// make mkfs executable
-	err = os.Chmod(Mkfs, 0775)
-	if err != nil {
+	if err := os.Chmod(mkfsFilePath, 0775); err != nil {
 		return errors.Wrap(err, 1)
 	}
 
-	if _, err = os.Stat(BootImg); os.IsNotExist(err) || force {
-		if err = DownloadFile(BootImg, fmt.Sprintf(baseUrl, "boot.img"), 600); err != nil {
+	if _, err := os.Stat(bootImgFilePath); os.IsNotExist(err) || force {
+		if err = DownloadFile(bootImgFilePath, fmt.Sprintf(baseUrl, "boot.img"), 600); err != nil {
 			return errors.Wrap(err, 1)
 		}
 	}
 
-	if _, err = os.Stat(KernelImg); os.IsNotExist(err) || force {
-		if err = DownloadFile(KernelImg, fmt.Sprintf(baseUrl, "stage3.img"), 600); err != nil {
+	if _, err := os.Stat(kernelImgFilePath); os.IsNotExist(err) || force {
+		if err = DownloadFile(kernelImgFilePath, fmt.Sprintf(baseUrl, "stage3.img"), 600); err != nil {
 			return errors.Wrap(err, 1)
 		}
 	}
 	return nil
 }
 
-func DownloadFile(filepath string, url string ,timeout int) error {
+func DownloadFile(filepath string, url string, timeout int) error {
 
 	fmt.Println("Downloading..", filepath)
 	out, err := os.Create(filepath + ".tmp")
 	if err != nil {
 		return err
 	}
-	defer out.Close()
-	
+	defer func() {
+		if err := out.Close(); err != nil {
```

Comment:
Probably no need to panic here, Close returns any errors from the underlying listener. I am fine with just out.Close and not checking for error here.

Reply:
Agree. I removed panics in defer functions but left as following to silence editor warnings like "Unhandled error..":
`defer func() {
		_ = out.Close()
	}()`
and
`defer func() {
		_ = resp.Body.Close()
	}()`

Reply:
yeh i mean those warnings are there for a reason - i'd prefer we handle all errors correctly - it's a go idiom (and just good programming as well)

Reply:
and you're right too.. there are no panics in that func but are returns.. so should I update that to `return` if `Close` returns error?

Reply:
Editors are just dump, they don't have any contextual information. So I won't worry too much about some editor plugin warning about it. But anyways just return the error and yes never panic inside the exported function. Let the caller handle it.

Reply:
Got it, updating PR..)

Reply:
I updated PR with reworked defer funcs